### ISSUE TITLE
Involution layer is not supported

### ISSUE THREAD

**Issue by SzaremehrjardiMT** [onnx/tensorflow-onnx#2053]
Involution layers

 Convolution has been the basis of most modern neural networks for computer vision. A convolution kernel is spatial-agnostic and channel-specific. Because of this, it isn't able to adapt to different visual patterns with respect to different spatial locations. Along with location-related problems, the receptive field of convolution creates challenges with regard to capturing long-range spatial interactions.
To address the above issues, Li et. al. rethink the properties
of convolution. The authors propose the "involution kernel", that is location-specific and channel-agnostic. Due to the location-specific nature of the operation, the authors say that self-attention falls under the design paradigm of
involution. [Ref](https://github.com/keras-team/keras-io/blob/master/examples/vision/involution.py)

I need to convert involution layers to onnx format. but it is not supported by tf2onnx and I couldn't find a good example to create a custom op using [--custom-ops ability. ](https://github.com/onnx/tensorflow-onnx/blob/v1.8.3/README.md#--custom-ops).

 involution layer source code can be found [here](https://github.com/keras-team/keras-io/blob/master/examples/vision/involution.py).

**Comment by SzaremehrjardiMT** [comment]
Thank you for your reply.
I couldn't use that example because I have Keras layers in my class so I don't know how to  make that tf op.

This is my custom layer:
Layer 1
x, _ = Involution(channel=3, group_number=1, kernel_size=3, stride=1, reduction_ratio=2, name="inv_rgb")(input_image, dep_image)
it is created by [Involution class.](https://github.com/keras-team/keras-io/blob/master/examples/vision/involution.py).

This is my tf op:

@tf_op("ExtractImagePatches")
class ExtractImagePatches:
    @classmethod
    def version_1(cls, ctx, node, **kwargs):
        node.type = "Involution"
        # Don't forget to set the domain!
        node.domain = constants.CONTRIB_OPS_DOMAIN
        # Ops defined using the custom ops framework only get access to inputs, not attributes
        
        
# Generate the kernel with respect to the input tensor.
        # B, H, W, K*K*G
        kernel_input = keras.layers.AveragePooling2D(pool_size=stride, strides=stride, padding="same")
        kernel = keras.Sequential([keras.layers.Conv2D(filters=channel // reduction_ratio, kernel_size=1),
                keras.layers.BatchNormalization(),
                keras.layers.ReLU(),
                keras.layers.Conv2D(filters=kernel_size * kernel_size * group_number,kernel_size=1,),])
        
        kernel_gen(kernel_input)

        # reshape the kerenl
        # B, H, W, K*K, 1, G
        kernel = kernel_reshape(kernel)

        # Extract input patches.
        # B, H, W, K*K*C
        input_patches = tf.image.extract_patches(
            images=x,
            sizes=[1, self.kernel_size, self.kernel_size, 1],
            strides=[1, self.stride, self.stride, 1],
            rates=[1, 1, 1, 1],
            padding="SAME",
        )

        # Reshape the input patches to align with later operations.
        # B, H, W, K*K, C//G, G
        input_patches = input_patches_reshape(input_patches)

        # Compute the multiply-add operation of kernels and patches.
        # B, H, W, K*K, C//G, G
        output = tf.multiply(kernel, input_patches)
        # B, H, W, C//G, G
        output = tf.reduce_sum(output, axis=3)

        # Reshape the output kernel.
        # B, H, W, C
        output = self.output_reshape(output)

        # Return the output tensor and the kernel.
        return output, kernel


I don't know how to access to stride, channel, group_number and reduction_ratio (the attributes of the layer) in tf op. How should I pass them?
I'm trying to follow this [tutorial](https://github.com/microsoft/onnxruntime-extensions/blob/main/tutorials/tf2onnx_custom_ops_tutorial.ipynb).

Thank you so much I really need a hint.

**Comment by fatcat-z** [comment]
I've gone through the source code you shared above.
I think for Involution class, it is more like a model than tf op, so probably we should not add a new custom op.
Let's just treat it as a model, just like a test [test_masking_custom](https://github.com/onnx/tensorflow-onnx/blob/c60c5eea2b9fc54b9b70c831e8f2953ca589bf89/tests/keras2onnx_unit_tests/test_layers.py#L2550).

Just process Involution class like MyPoolingMask in the above test, tf2onnx should be able to handle it.

**Comment by SzaremehrjardiMT** [comment]
Thank you! I changed my custom layers a little now they don't use "tf.image.extract_patches". so, I could convert my model to onnx.

**Comment by fatcat-z** [comment]
> Thank you! I changed my custom layers a little now they don't use "tf.image.extract_patches". so, I could convert my model to onnx.

Good to hear this!
